src/multi_vector_simulator/E4_verification.py

```diff
@@ -154,3 +156,50 @@ def detect_excessive_excess_generation_in_bus(dict_values):
             if ratio < 0.9:
                 msg = f"Attention, on bus {bus_label} there is excessive excess generation, totalling up to {round((1 - ratio) * 100)}% of the inflows. The total inflows are {round(total_inflow_bus)} and outflows {round(total_outflow_bus)}  It seems to be cheaper to have this excess generation than to install more capacities that forward the energy carrier to other busses (if those assets can be optimized)."
                 logging.warning(msg)
+
+
+def verify_state_of_charge(dict_values):
+    r"""
+    This function checks the state of charge of each storage component
+    It raises warning log messages if the SoC has a physically infeasible value
+
+    Parameters
+    ----------
+    dict_values
+
+    Returns
+    -------
+    - Nothing if there are no physically infeasible SoC values for the storage components
+    - Prints log messages to console and log file if there are physically impossible SoC values
+
+    Notes
+    -----
+    Tested with:
+    - test_E4_verification.test_verify_state_of_charge_feasible()
+    - test_E4_verification.test_verify_state_of_charge_soc_below_zero()
+    - test_E4_verification.test_verify_state_of_charge_soc_above_zero()
+
+    """
+    # Dict holding the data of all of the storage components
+    storage_data_dict = dict_values[ENERGY_STORAGE]
+
+    # Check if storage components are present in the energy system and only then verify SoC values
+    if bool(storage_data_dict) is not False:
+        # Loop through the storage components
+        for storage in list(storage_data_dict.keys()):
```

Comment:
This is a new way for me to loop though the components - I would probably have used

```
for storage in dict_values[ENERGY_STORAGE]:
    soc_timeseries = dict_values[ENERGY_STORAGE][storage][TIMESERIES_SOC] 
```
As you wrote tests, your code also is valid.

Reply:
I just used that if test to not execute the code in the for loop in case there is no storage component at all.